File: src/main/kotlin/org/skialin/SVGDOM.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

/** A parsed SVG document. Mirrors Skia's `SkSVGDOM`. */
class SVGDOM private constructor(
    ptr: Long,
) : Managed(ptr, SVGDOMNative::nRelease) {
    /** The viewport used to resolve the root's width/height when specified in relative units. */
    var containerSize: Pair<Float, Float>
        get() = SVGDOMNative.nGetContainerSize(nativePtr).let { it[0] to it[1] }
        set(value) = SVGDOMNative.nSetContainerSize(nativePtr, value.first, value.second)

    /**
     * Sets the container size AND forcibly overrides the root `<svg>`'s width/height/
     * preserveAspectRatio so the document always stretches to fill (width, height) on render,
     * regardless of its own intrinsic size or viewBox -- unlike [containerSize] alone, which only
     * affects percentage-valued root width/height, not absolute ones (e.g. `width="24"`).
     */
    fun setSizeAndStretch(
        width: Float,
        height: Float,
    ) = SVGDOMNative.nSetSizeAndStretch(nativePtr, width, height)

    fun render(canvas: Canvas) = SVGDOMNative.nRender(nativePtr, canvas.ptr)

    companion object {
        /** Parses [bytes] as SVG XML. `null` if it doesn't parse. */
        fun makeFromBytes(bytes: ByteArray): SVGDOM? = SVGDOMNative.nMakeFromBytes(bytes).takeIf { it != 0L }?.let { SVGDOM(it) }
    }
}

private object SVGDOMNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nMakeFromBytes(bytes: ByteArray): Long

    external fun nRelease(ptr: Long)

    external fun nSetContainerSize(
        ptr: Long,
        width: Float,
        height: Float,
    )

    external fun nGetContainerSize(ptr: Long): FloatArray

    external fun nSetSizeAndStretch(
        ptr: Long,
        width: Float,
        height: Float,
    )

    external fun nRender(
        ptr: Long,
        canvasPtr: Long,
    )
}

File: rust/skialin-core/src/svg_dom.rs
use crate::{sys, Canvas};

/// A parsed SVG document. Mirrors Skia's `SkSVGDOM`.
pub struct SVGDOM(*mut sys::SkSVGDOM);

impl SVGDOM {
    /// Parses `bytes` as SVG XML. `None` if it doesn't parse.
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        let ptr = unsafe { sys::skialin_bridge_SVGDOM_MakeFromStream(bytes.as_ptr(), bytes.len()) };
        (!ptr.is_null()).then_some(SVGDOM(ptr))
    }

    /// Sets the viewport used to resolve the root's width/height when
    /// they're specified in relative units.
    pub fn set_container_size(&mut self, width: f32, height: f32) {
        unsafe { sys::skialin_bridge_SVGDOM_setContainerSize(self.0, width, height) };
    }

    pub fn container_size(&self) -> (f32, f32) {
        let mut width = 0.0;
        let mut height = 0.0;
        unsafe { sys::skialin_bridge_SVGDOM_getContainerSize(self.0, &mut width, &mut height) };
        (width, height)
    }

    /// Sets the container size AND forcibly overrides the root's width/height/preserveAspectRatio
    /// so the document always stretches to fill (width, height) on render, regardless of its own
    /// intrinsic size or viewBox. Unlike `set_container_size`, this affects absolute-unit root
    /// width/height too, not just percentage ones.
    pub fn set_size_and_stretch(&mut self, width: f32, height: f32) {
        unsafe { sys::skialin_bridge_SVGDOM_setSizeAndStretch(self.0, width, height) };
    }

    pub fn render(&self, canvas: &mut Canvas) {
        unsafe { sys::skialin_bridge_SVGDOM_render(self.0, canvas.as_raw()) };
    }
}

impl Drop for SVGDOM {
    fn drop(&mut self) {
        unsafe { sys::skialin_bridge_SVGDOM_unref(self.0) };
    }
}
